### PR TITLE
document, set, and string distance metrics

### PR DESCRIPTION
**Pull request by bdewilde**
Changes:

- Added a `distance.py` module containing several document, set, and string distance metrics
    - word movers distance: document distance as distance between individual words represented by word2vec vectors, normalized
    - "word2vec" distance: token, span, or document distance as cosine distance between (average) word2vec representations, normalized
    - jaccard distance: string or set(string) distance as intersection / overlap, normalized, with optional fuzzy-matching across set members
    - hamming: distance between two strings as number of substititions, optionally normalized
    - levenshtein: distance between two strings as number of substitions, deletions, and insertions, optionally normalized
    - jaro-winkler: distance between two strings with variable prefix weighting, normalized

TODO:
- add tests! soon...

@danvalente @jeiranj -- care to take a look?

### CODE REVIEW

File: textacy/distance.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+from sklearn.metrics import pairwise_distances
+from spacy.strings import StringStore
+
+from textacy import extract
+from textacy.compat import str
+
+
+def word_movers_distance(doc1, doc2, metric='cosine'):
+    """
+    Measure the semantic distance between two documents using Word Movers Distance.
```
Comment (danvalente) [inline textacy/distance.py]: Ref or explanation for what Word Movers Distance is?

Comment (bdewilde) [inline textacy/distance.py]: http://jmlr.org/proceedings/papers/v37/kusnerb15.pdf

Comment (danvalente) [inline textacy/distance.py]: No, I mean, put it somewhere in the docstring.

File: textacy/distance.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    Args:
+        doc1 (`TextDoc` or `spacy.Doc`)
+        doc2 (`TextDoc` or `.spacy.Doc`)
```
Comment (jeiranj) [inline textacy/distance.py]: doc2 (`TextDoc` or `spacy.Doc`), you meant, correct?  Without the dot.

Comment (bdewilde) [inline textacy/distance.py]: ha, good eye ;)

File: textacy/distance.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                n += 1
+    distance_mat = pairwise_distances(np.array(word_vecs), metric=metric).astype(np.double)
+    distance_mat /= distance_mat.max()
+
+    vec1 = collections.Counter(
+        stringstore[word.text] - 1
+        for word in extract.words(doc1)
+        if word.has_vector)
+    vec1 = np.array([vec1[word_idx] for word_idx in range(len(stringstore))]).astype(np.double)
+    vec1 /= vec1.sum()  # normalize word counts
```
Comment (jeiranj) [inline textacy/distance.py]: only l1 normalization?  How about someone wants l2 normalization?  Just a thought.

File: textacy/distance.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+    vec2 = collections.Counter(
+        stringstore[word.text] - 1
+        for word in extract.words(doc2)
+        if word.has_vector)
+    vec2 = np.array([vec2[word_idx] for word_idx in range(len(stringstore))]).astype(np.double)
+    vec2 /= vec2.sum()  # normalize word counts
```
Comment (jeiranj) [inline textacy/distance.py]: same here.

File: textacy/texts.py [hunk 5]
```diff
@@ -643,7 +643,10 @@ def add_text(self, text, lang=None, metadata=None):
         doc.corpus = self
         self.docs.append(doc)
         self.n_docs += 1
-        self.n_sents += doc.n_sents
+        try:
+            self.n_sents += doc.n_sents
+        except ValueError:  # no parser loaded, skip it
+            pass
         self.n_tokens += doc.n_tokens
```
Comment (jeiranj) [inline textacy/texts.py]: shouldnt this line go in the try except block too?

Comment (bdewilde) [inline textacy/texts.py]: nope! sentence segmentation (for better or worse) relies on spacy's syntactic parse, while tokenization does not.

**Pull request merged**